pkg/controller/clustercatalog_controller.go

```diff
@@ -0,0 +1,156 @@
+package controller
+
+import (
+	"fmt"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/client-go/config/clientset/versioned/scheme"
+	"github.com/openshift/cluster-olm-operator/pkg/clients"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/management"
+	"golang.org/x/net/context"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+
+	catalogdv1alpha1 "github.com/operator-framework/catalogd/api/core/v1alpha1"
+)
+
+func NewClusterCatalogController(name string, manifest []byte, key types.NamespacedName, operatorClient *clients.OperatorClient, dynamicClient dynamic.Interface, clusterCatalogClient *clients.ClusterCatalogClient, recorder events.Recorder) factory.Controller {
+	c := &clusterCatalogController{
+		manifest:         manifest,
+		name:             name,
+		key:              key,
+		applyFunc:        defaultApplyFunc(dynamicClient),
+		managedFunc:      defaultManagedFunc(operatorClient),
+		shouldUpdateFunc: defaultShouldUpdateFunc(),
+		catalogGetFunc:   clusterCatalogClient.Get,
+	}
+
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).WithInformers(operatorClient.Informer(), clusterCatalogClient.Informer()).ToController(c.name, recorder)
+}
+
+func defaultApplyFunc(client dynamic.Interface) applyFunc {
+	return func(ctx context.Context, key types.NamespacedName, fieldManager string, force bool, gvr schema.GroupVersionResource, manifest []byte) error {
+		var resourceInterface dynamic.ResourceInterface = client.Resource(gvr)
+		if key.Namespace != "" {
+			resourceInterface = client.Resource(gvr).Namespace(key.Namespace)
+		}
+		_, err := resourceInterface.Patch(
+			ctx,
+			key.Name,
+			types.ApplyPatchType,
+			manifest,
+			metav1.PatchOptions{
+				Force:        ptr.To(force),
+				FieldManager: fieldManager,
+			})
+		return err
+	}
+}
+
+func defaultManagedFunc(oc *clients.OperatorClient) managedFunc {
+	return func() (bool, error) {
+		operatorSpec, _, _, err := oc.GetOperatorState()
+		if err != nil {
+			return false, err
+		}
+		return management.IsOperatorManaged(operatorSpec.ManagementState) && (operatorSpec.ManagementState != operatorv1.Removed || management.IsOperatorNotRemovable()), nil
+	}
+}
+
+func defaultShouldUpdateFunc() shouldUpdateFunc {
+	return func(manifest []byte, u *unstructured.Unstructured) (bool, error) {
+		required, _, err := scheme.Codecs.UniversalDecoder().Decode(manifest, nil, &unstructured.Unstructured{})
```

Comment:
I'm having trouble groking the unstructured objects.
1. Can you give `u` a more descriptive name?
2. Maybe make a variable (or add a comment) for the `&unstructured.Unstructured{}` that we pass to `Decode` so that it is more obvious what that parameter is for.

Reply:
I can give `u` a more descriptive name. 

For point 2, see https://pkg.go.dev/k8s.io/apimachinery@v0.31.1/pkg/runtime#Decoder for more information on the inputs and outputs. The TLDR is that the `&unstructured.Unstructured{}` is the target type for the returned `runtime.Object` and may or may not be populated. I'll add a comment with more information.

Reply:
Updated in https://github.com/openshift/cluster-olm-operator/pull/73/commits/5b237d4e4d462f19ff7571661b839393bde6b676